File: 98. Validate Binary Search Tree/solution.go
func isValidBST(root *TreeNode) bool {
    if root == nil {
        return true
    }
    
    // Get the inorder traversal
    list := inorder(root)
    
    // Check if the elements in the list are sorted in increasing order
    for i:=0; i<len(list)-1;i++{
        if list[i] >= list[i+1] {
            return false
        }
    }
    return true
}

func inorder(root *TreeNode) []int {
    if root.Left == nil && root.Right == nil {
        return []int{root.Val}
    } else if root.Left != nil && root.Right == nil {
        left := inorder(root.Left)
        return append(left, root.Val)
    } else if root.Left == nil && root.Right != nil {
        right := inorder(root.Right)
        return append([]int{root.Val}, right...)
    } else {
        left := inorder(root.Left)
        right := inorder(root.Right)
        return append(left, append([]int{root.Val}, right...)...)
    } 
}

File: 98. Validate Binary Search Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, null, null);
    }
    
    public boolean isValidBST(TreeNode root, Integer left, Integer right) {
        if (root == null) {
            return true;
        }
        
        boolean isValid = true;
        
        if ((right != null) && (root.val >= right)) 
            isValid = false;
        if (isValid && (left != null) && (root.val <= left))
            isValid = false;
        if (isValid)
            return isValidBST(root.left, left, root.val) && isValidBST(root.right, root.val, right);
        else
            return false;
    }
}
